patroni/validator.py

```diff
@@ -0,0 +1,230 @@
+            elif callable(schema):
+                try:
+                    schema(data)
+                    return True
+                except Exception as e:
+                    if not quiet:
+                        logger.warning("%s ('%s') didn't pass validation: %s", self.get_fullpath(key), data, e)
+            elif isinstance(schema, Or):
+                counter = 0
+                for v in schema.args:
+                    counter += (1 if self.validator(key, data, v, quiet=True) else 0)
+                if counter < 1:
+                    logger.warning("%s ('%s') didn't pass validation", self.get_fullpath(key), data)
+                else:
+                    return True
+
+            else:
+                if not quiet:
+                    logger.warning("%s type is not %s", self.get_fullpath(key), _get_type_name(type(schema)))
+        elif isinstance(data, dict):
+            Schema(schema, path=self.get_fullpath(key))(data)
+        elif isinstance(data, list):
+            Schema(schema, path=self.get_fullpath(key))(data)
+
+    def validate_orcase(self, key, schema, data):
+        counter = 0
+        for a in key.args:
+            if a in data:
+                counter += 1
+                subschema = schema._schema[a]
+                self.validator(a, data[a], subschema)
+        if counter < 1:
+            logger.warning("neither of %s is defined", ", ".join([self.get_fullpath(a) for a in key.args]))
+
+    def get_fullpath(self, path):
+        return self._path + ("." if self._path else "") + str(path)
+
+
+def _get_type_name(python_type):
+    if python_type == str:
+        return "string"
+    elif python_type == int:
+        return "integer"
+    elif python_type == float:
+        return "number"
+    elif python_type == bool:
+        return "boolean"
+    elif python_type == list:
+        return "array"
+    elif python_type == dict:
+        return "dictionary"
+    return "string"
+
+
+userattributes = {"username": "", Optional("password"): ""}
+available_dcs = [m.split(".")[-1] for m in dcs_modules()]
+
+schema = Schema({
+  "name": str,
+  "scope": str,
+  "restapi": {
+    "listen": validate_host_port
+  },
+  Optional("bootstrap"): {
+    "dcs": {
+        Optional("ttl"): int,
+        Optional("loop_wait"): int,
+        Optional("retry_timeout"): int,
+        Optional("maximum_lag_on_failover"): int
+        },
+    "pg_hba": [str],
+    "initdb": [Or(str, dict)]
+  },
+  Or(*available_dcs): Case({
+      "etcd": {
+          Or("host", "hosts", "srv", "url", "proxy"): Case({
+              "host": str,
+              "hosts": Or(str, list),
+              "srv": str,
+              "url": str,
+              "proxy": str})
+         },
+      "exhibitor": {
+          "hosts": str,
+          },
+      "zookeeper": {
+          "hosts": [validate_host_port],
+          },
+      "kubernetes": {
```

Comment:
There are a bit more parameters for kubernetes: https://github.com/zalando/patroni/blob/master/docs/SETTINGS.rst#kubernetes

Most of them are optional, but the structure is not really obvious.